### ISSUE TITLE
`record()` turns everything optional after updating package

### ISSUE THREAD

**Issue by Addeuz** [tskj/typescript-json-decoder#17]
Hello,

I've updated the package from `v1.0.6` to `v1.0.10` to get the new optional features (types gets ?).

But now I've running into the issue that using the `record` decoder turns every key optional..
If I create this simple decoder:
```ts
type Admin = decodeType<typeof adminDecoder>;
const adminDecoder = record({
  id: string,
  name: string,
  rank: optional(number)
});

const adminVar: Admin = {};
```
and then in VSCode hover `adminDecoder` I get this:
```ts
const adminDecoder: DecoderFunction<{
    id?: string;
    name?: string;
    rank?: number;
}>
```
If I hover the type Admin I get this:
```ts
type Admin = {
    id?: string;
    name?: string;
    rank?: number;
}
```
And I can create the variable `admin` like that without getting any ts errors..

This is in an old project at my job.

But whats really disturbing is that if I create a new ts project and install the package. Everything works as expected....
This code:
```ts
const adminDecoder = record({
  id: string,
  name: string,
  rank: optional(string)
});
```
gives me:
```ts
const adminDecoder: DecoderFunction<{
    id: string;
    name: string;
    rank?: string | undefined;
}>
```
As expected...

My ts version is the same in the two projects.
And I've tried to uninstall and install this package..

I've cleared the cache in vscode as well. So I'm not sure what's going on...

EDIT: Reverting the package to `v1.0.6` makes it so it goes back to the same as it was before... The variable `admin` gives me the error that `id`, `name` and `rank` is missing..

**Comment by tskj** [comment]
This is the second issue on this today! Sorry about this, this is definitely a bug. Trying to get the `?` feature working was a real hassle, and the 1.0.10 release was a workaround to a bug in the TypeScript type inference itself, related to this feature. So I'm not sure what can be done about this, but I'll have a look.

You might try 1.0.9, or some other version inbetween, if you don't need the particular feature / struggle with the particular bug that this workaround was intended to fix. 

Please share more examples of things that do or don't work on the two versions, and I'll add them to the test cases for better coverage in the future.

**Comment by Addeuz** [comment]
When I go to 1.0.8 from it breaks again.. But the weird thing is that when I hover the `record` method it shows me this:
```ts
(alias) record<{
    id: DecoderFunction<string>;
    name: DecoderFunction<string>;
    rank: DecoderFunction<number>;
}>(s: {
    id: DecoderFunction<string>;
    name: DecoderFunction<string>;
    rank: DecoderFunction<number>;
}): DecoderFunction<...>
import record
```
Which looks correct to me.
But then when I hover the variable it shows that everything is optional...

So it seems like something is broken in my setup. Since it works on a new project?

**Comment by Addeuz** [comment]
I fixed the error. Turning on `strict: true` in my tsconfig fixed it. Unfortunately this broke a lot of my custom decoders. Sooo, know I'll have to learn how to write proper decoders for this haha. But should be fine

**Comment by tskj** [comment]
Good to hear! Did turning on `strict: true` fix 1.0.10 as well, or an earlier version? 

Should probably add a note about `strict: true` to the readme, because I've definitely developed the entire project and only used it with that setting!

**Comment by Addeuz** [comment]
Yes, `strict: true` fixed it for 1.0.10 as well :smile:
